### PR TITLE
Fix Tab Completion Context Detection

### LINKED ISSUE #14836: Tab completion for globals does not work in lines with dots
To reproduce:

```python
import numpy as np
long_variable_name = 56

%config Completer.use_jedi = False

x = np.random(lo   # pressing <tab> here does not show `long_variable_name`
```

| Last working | Broken |
|---|---|
|  |  |

https://github.com/ipython/ipython/compare/8.26.0...8.27.0

The problem is a heuristic restored in https://github.com/ipython/ipython/pull/14474, which checks for `"."` in the line to decide between using attribute and global matcher, which is confirmed by the following working correctly:

Image

### PR DESCRIPTION
**Pull request by Darshan808**
### Fixes #14836 

This PR resolves the issue with IPython tab completion incorrectly triggering attribute completion when a dot (`.`) appears in the line. Instead of just checking for '.', I've added a new function, `_determine_completion_context`, to accurately determine whether the cursor is in an attribute or global completion context. 

The solution has been tested against multiple edge cases (e.g., nested expressions, strings, comments) and handles them all correctly. Please review the changes and let me know if you have any feedback or suggestions!

### CODE REVIEW

**Comment by krassowski** [comment]
I think it a good idea to have both unit tests (for `_determine_completion_context`) and a few integration tests. Since https://github.com/ipython/ipython/pull/14837 already includes one integration test, you could build on that but we would not want to add too many.

**Comment by Darshan808** [comment]
I’ve added two more integration tests: one for completion inside f-strings and another for completion inside dictionaries. I believe these cover the main cases, but do you think there are any other scenarios worth testing?

Also, regarding the test failures, I see that one test, test_greedy_completions, is consistently failing. I’m looking into it, but I haven’t found any reasons for why it’s failing yet.

File: tests/test_completer.py [hunk 1]
```diff
@@ -1747,6 +1785,47 @@ def _(expected):
             _(["completion_a"])
 
 
+@pytest.mark.parametrize(
+    "line,expected",
+    [
+        # Basic test cases
+        ("np.ran", "attribute"),
```
Comment (krassowski) [inline tests/test_completer.py:1792]: Does this work?

```suggestion
        ("a[0].r", "attribute"),
```

Comment (Darshan808) [reply tests/test_completer.py:1792]: Just found the issue

**Comment by Darshan808** [comment]
I modified the regex to check for patterns like `<anything>.w`. I'm not entirely sure if this is the best approach, but it seems to work well and passes all the test cases I could think of.

File: IPython/core/completer.py [hunk 2]
```diff
@@ -2360,11 +2360,118 @@ def _jedi_matches(
...
+            return self._CompletionContextType.ATTRIBUTE
+
+        return self._CompletionContextType.GLOBAL
+
+    def _is_in_string_or_comment(self, text):
+        # Check for comments
+        if "#" in text:
+            parts = text.split("#")
+            if len(parts) > 1:
+                return True
```
Comment (krassowski) [inline IPython/core/completer.py:2392]: 1. What is the `split()` used for?
2. What about `d["#"].a` - this will be relatively common with pandas column access, some columns will contain `#` in name

File: IPython/core/completer.py [hunk 3]
```diff
@@ -2360,11 +2360,118 @@ def _jedi_matches(
...
+        in_single_quote = False
+        in_double_quote = False
+        in_triple_single = False
+        in_triple_double = False
+        in_fstring = False
+        in_fstring_expression = False
+        i = 0
+
+        while i < len(text):
+            # Check for f-string start
```
Comment (krassowski) [inline IPython/core/completer.py:2403]: To be future-proof we could allow t-strings (https://peps.python.org/pep-0750/). There is a strong indication that this JEP will be approved (https://discuss.python.org/t/pep750-template-strings-new-updates/71594/118).

File: IPython/core/completer.py [hunk 4]
```diff
@@ -2360,11 +2360,118 @@ def _jedi_matches(
...
+    def _determine_completion_context(self, line):
+        """
+        Determine whether the cursor is in an attribute or global completion context.
+        """
+        if self._is_in_string_or_comment(line):
+            return self._CompletionContextType.GLOBAL
+        if line.endswith("."):
+            return self._CompletionContextType.ATTRIBUTE
+
+        last_token_match = re.search(r"([\w]+)$", line)
```
Comment (krassowski) [inline IPython/core/completer.py:2376]: Can you add a comment with an example explaining which case it addresses?

In Python tokens can include digits (but cannot start with digits), not sure if this is relevant here yet.

File: IPython/core/completer.py (hunk 2 above)
Comment (Darshan808) [reply IPython/core/completer.py]: Improved the logic for comment checking!

File: IPython/core/completer.py (hunk 3 above)
Comment (Darshan808) [reply IPython/core/completer.py]: Thanks. Added this!

File: IPython/core/completer.py (hunk 4 above)
Comment (Darshan808) [reply IPython/core/completer.py]: Added comments!
>  can include digits (but cannot start with digits)

Initially, I didn't think this was important, but I've now updated the regex to account for it.

**Comment by Darshan808** [comment]
After reanalyzing the logic, I found redundant checks and simplified it to a single regex search.

File: IPython/core/completer.py [hunk 5]
```diff
@@ -2360,11 +2360,122 @@ def _jedi_matches(
...
+        return (
+            in_single_quote or in_double_quote or in_triple_single or in_triple_double
+        )
+
     @context_matcher()
     def python_matcher(self, context: CompletionContext) -> SimpleMatcherResult:
         """Match attributes or global python names"""
         text = context.line_with_cursor
-        if "." in text:
+        completion_type = self._determine_completion_context(text)
```
Comment (krassowski) [inline IPython/core/completer.py:2477]: A fundamental problem here is that it ignores the cursor position. So for example this works:

image

This fails:

image

I think we need to use `context.text_until_cursor` instead, see:

https://github.com/ipython/ipython/blob/801c9adb94b213fc5e8cacb0d9920a0ff7727c2e/IPython/core/completer.py#L618-L653

File: IPython/core/completer.py [hunk 6]
```diff
@@ -2360,11 +2360,122 @@ def _jedi_matches(
...
+        i = 0
+
+        while i < len(text):
+            # Check for f-string or t-string start
+            if (
+                i + 1 < len(text)
+                and text[i] in ("f", "t")
+                and (text[i + 1] == '"' or text[i + 1] == "'")
+            ):
+                in_template_string = True
```
Comment (krassowski) [inline IPython/core/completer.py:2401]: This is never reset to `False`, here is how it fails - can you add a test case?

image

File: IPython/core/completer.py [hunk 7]
```diff
@@ -2360,11 +2360,122 @@ def _jedi_matches(
...
+        # Ex: 'obj.', 'np.random.ran'.
+        chain_match = re.search(r"([a-zA-Z_].*)\.([a-zA-Z_][a-zA-Z0-9_]*)?$", line)
+        if chain_match:
+            return self._CompletionContextType.ATTRIBUTE
+
+        # No attribute pattern → GLOBAL.
+        # Ex: 'var', '', '3.', '-42.5.'
+        return self._CompletionContextType.GLOBAL
+
+    def _is_in_string_or_comment(self, text):
```
Comment (krassowski) [inline IPython/core/completer.py:2385]: This will fail on nested f-strings:

image

I am not sure if we want to support it (if it is easy then sure, if it is hard - not sure), but at least need to document it, possibly by adding xfail test?

File: IPython/core/completer.py [hunk 8]
```diff
@@ -2360,11 +2360,122 @@ def _jedi_matches(
...
+
+        # Match 'anything-dot-word' at end for attribute context.
+        # Ex: 'obj.', 'np.random.ran'.
+        chain_match = re.search(r"([a-zA-Z_].*)\.([a-zA-Z_][a-zA-Z0-9_]*)?$", line)
+        if chain_match:
+            return self._CompletionContextType.ATTRIBUTE
+
+        # No attribute pattern → GLOBAL.
+        # Ex: 'var', '', '3.', '-42.5.'
+        return self._CompletionContextType.GLOBAL
```
Comment (krassowski) [inline IPython/core/completer.py:2383]: Some edge cases not covered, probably not important enough to cover, but listing them for completness:

```
(3).to_<tab>
```

expecting `(3).to_bytes`

Note that `3.to_bytes` is invalid in Python

```
3.1.as_<tab>
-3.1.r<tab>
```

Expecting `3.1.as_integer_ratio` and `-3.1.real` respectively.

**Review by krassowski — COMMENTED** [review]
Just a few edge cases

**Comment by Darshan808** [comment]
I’ve run into an interesting issue with the python_matcher. For `f'{a.`, it correctly suggests attributes like .append (context_type is attribute). But for f'{a.}' with the cursor in the same spot, despite also identifying as ATTRIBUTE, it completes to .ipynb instead.
```python
  # Debug line
   completion_type = self._determine_completion_context(text)
   print(completion_type, " for ", text)
```
```bash
a = []

In [3]: f'{a.|        # suggests .append
_CompletionContextType.ATTRIBUTE  for  f'{a.

In [4]: f'{a.|}'     # suggests .ipynb
_CompletionContextType.ATTRIBUTE  for  f'{a.
```

**Comment by Darshan808** [comment]
> Maybe it is because it uses the line_buffer instead of text_until_cursor?

Yes, using argument `text` instead of `text_until_cursor` fixed the issue. Thanks!

File: IPython/core/completer.py [hunk 9]
```diff
@@ -2360,11 +2360,213 @@ def _jedi_matches(
...
+            if expr_start >= 0:
+                # We're looking at the expression inside a template string
+                expr = line[expr_start + 1 :]
+                # Recursively determine the context of the expression
+                return self._determine_completion_context(expr)
+
+        # Match tuples followed by a dot (e.g., (a, b).index)
+        tuple_attr_match = re.search(
+            r"\(\s*[a-zA-Z_][a-zA-Z0-9_]*\s*(,\s*[a-zA-Z_][a-zA-Z0-9_]*\s*)+\)\.$", line
+        )
```
Comment (krassowski) [inline IPython/core/completer.py:2389]: This is overly strict as it only handles tuples with two or more identifiers. It will miss on cases like:
- `(a,)`
- `(1, 2)`
- `('a', 'b')`
etc

Now, I don't know if we should special-case tuples over other similar collections (lists, sets). I also worry that making this syntax more complex to account for all these cases would make it slower, and completer needs to be very fast.

I know that using `tokenizer` out right will often not work as it does not work for incomplete lines, but could we use a hybrid approach where for cases like this - where we assume that there is a valid tuple (list, set) followed by a dot we just remove the final dot and see if tokenizer can parse it and if the result is one of: list, tuple, dictionary, set?

File: IPython/core/completer.py [hunk 10]
```diff
@@ -2360,11 +2360,213 @@ def _jedi_matches(
             else:
                 return iter([])
 
+    class _CompletionContextType(enum.Enum):
+        ATTRIBUTE = "attribute"  # For attribute completion
+        GLOBAL = "global"  # For global completion
+
+    def _determine_completion_context(self, line):
```
[the remark this answers is not in the document]
Comment (krassowski) [reply IPython/core/completer.py:2367]: I think fixing this one might be of relatively little value.

File: IPython/core/completer.py (hunk 9 above)
Comment (fleming79) [reply IPython/core/completer.py:2389]: If you're interested in speed you could compile the expression to speed things up.

https://docs.python.org/3/library/re.html#re.compile

**Comment by Darshan808** [comment]
There are so many edge cases, and trying to make the logic perfect keeps causing some tests to fail. Covering all cases would require a lot of extra checks. My initial approach of matching `anything.word` handles most cases well, with only a few minor flaws. If we don’t need to be too strict, this simpler solution with just two regex expressions could work. What do you think?
I have added unsupported checks in xfail tests.

**Review by krassowski — APPROVED** [review]
Thank you @Darshan808. I think this is a good trade-off between completeness and complexity. I left one tiny suggestion on making part of the code easier to read.

**Comment by krassowski** [comment]
@meeseeksdev please backport to 8.x
